app/models/mixins/container_deployment_mixin.rb

```diff
@@ -0,0 +1,211 @@
+      [unreachable_ips.empty?, unreachable_ips]
+    end
+    raise StandardError, "couldn't connect to : #{unreachable_ips.join(',')}" unless success
+    true
+  end
+
+  def subscribe_deployment_master
+    rhsub_user = rhsm_auth.userid
+    rhsub_pass = rhsm_auth.password
+    rhsm_sku = rhsm_auth.rhsm_sku
+    perform_agent_commands(["sudo subscription-manager register --username='#{rhsub_user}' --password='#{rhsub_pass}'"])
+    pool_cmdret = perform_agent_commands(["sudo subscription-manager list --available --matches='#{rhsm_sku}' --pool-only"])
+    pool_id = pool_cmdret[:stdout].split("\n").first.to_s.strip
+    perform_agent_commands(["sudo subscription-manager attach --pool='#{pool_id}'"]) unless pool_id.blank?
+    enabled_repos_cmdret = perform_agent_commands(['sudo subscription-manager repos --disable="*"',
+                                                   "sudo subscription-manager repos --enable=#{SUBSCRIPTION_REPOS.join(" --enable=")}",
+                                                   "sudo subscription-manager repos --list-enabled"])
+    enabled_repos = enabled_repos_cmdret[:stdout].split(" ").to_a.map(&:strip).reject(&:empty?)
+    missing_repos = SUBSCRIPTION_REPOS - enabled_repos
+    raise StandardError, "couldn't register to following repos : #{missing_repos.join(',')} with SKU : #{rhsm_sku}" unless missing_repos.empty?
+  end
+
+  def subscribe_cluster
+    perform_agent_commands(["sudo mv ~/#{RHEL_SUBSCRIPTION_FILE_NAME} #{INVENTORIES_PATH}"])
+    run_playbook_command("ansible-playbook #{INVENTORIES_PATH}playbooks/byo/rhel_subscribe.yml -i "\
+                         "#{INVENTORIES_PATH}#{RHEL_SUBSCRIPTION_FILE_NAME} 1> #{ANSIBLE_LOG} 2> #{ANSIBLE_ERROR_LOG} < /dev/null")
+  end
+
+  def provision_tasks
+    automation_task.automation_request.reload
+    return [] unless automation_task.automation_request.options["master_request_ids"]
+    ids = automation_task.miq_request.options["master_request_ids"] + automation_task.miq_request.options["node_request_ids"]
+    MiqRequest.where(:id => ids)
+  end
+
+  def provision_started?
+    provision_tasks.count > 0
+  end
+
+  def nodes_subscription_needed?
+    container_deployment_nodes.count > 1
+  end
+
+  def agent_exists?
+    agent_pid
+  end
+
+  def ssh_user
+    ssh_auth.userid
+  end
+
+  def add_automation_task(task)
+    send(:update_attributes!, :automation_task_id => task.id)
+  end
+
+  def customize(options = {})
+    self.customizations = {:agent => {}} if customizations.empty?
+    customizations[:agent].merge!(options)
+    save!
+  end
+
+  def perform_scp(local_path, remote_path)
+    ip = roles_addresses("deployment_master")
+    username = ssh_user
+    Net::SCP.upload!(ip, username, local_path, remote_path, :ssh => {:key_data => ssh_auth.auth_key})
+  end
+
+  def perform_agent_commands(commands = [])
+    ip = roles_addresses("deployment_master")
+    username = ssh_user
+    @ssh ||= LinuxAdmin::SSH.new(ip, username, ssh_auth.auth_key)
```

Comment:
To prevent doing extra work on subsequent calls, this should probably be:
```ruby
@ssh ||= begin
  ip = roles_addresses("deployment_master")
  username = ssh_user
  LinuxAdmin::SSH.new(ip, username, ssh_auth.auth_key)
end
```